git/repo.go

```diff
@@ -71,24 +72,29 @@ func (f optionFunc) apply(r *Repo) {
 	f(r)
 }
 
+// PollInterval is the period at which Flux checks for updated images
 type PollInterval time.Duration
 
 func (p PollInterval) apply(r *Repo) {
 	r.interval = time.Duration(p)
 }
 
+// Timeout is the duration of time after which git operations time out
 type Timeout time.Duration
 
 func (t Timeout) apply(r *Repo) {
 	r.timeout = time.Duration(t)
 }
 
-var ReadOnly optionFunc = func(r *Repo) {
-	r.readonly = true
+// RepoIsReadOnly sets the repo to be notated as being readonly
+type RepoIsReadOnly bool
```

Comment:
We discussed this on our call, I'm sorry if some wire got crossed.  The purpose for this change is prior to this commit there was no way to conditionally use this `ReadOnly` option, it was either hardcoded or not.  That likely results from the fact that prior to this PR there was no way for the user to configure any readonly modes.  Now that there is a way to configure it with a flag we have a design decision to make:

## Option A

If we follow the way the rest of the options look:
```go
// PollInterval is the period at which Flux checks for updated images
type PollInterval time.Duration

func (p PollInterval) apply(r *Repo) {
	r.interval = time.Duration(p)
}

// Timeout is the duration of time after which git operations time out
type Timeout time.Duration

func (t Timeout) apply(r *Repo) {
	r.timeout = time.Duration(t)
}
```

and write our `ReadOnly` option function to follow suit exactly:
```go
// RepoIsReadOnly sets the repo to be notated as being readonly
type RepoIsReadOnly bool

func (r RepoIsReadOnly) apply(repo *Repo) {
	repo.readonly = bool(r)
}
```

it allows us to write code that flows quite naturally when we're using a boolean set elsewhere to configure the option:
```go
	options := []git.Option{
		git.PollInterval(*gitPollInterval),
		git.Timeout(*gitTimeout),
		git.RepoIsReadOnly(*gitReadonly),
	}

	repo := git.NewRepo(gitRemote, options...)
```

## Option B

however.. if we stray from that model and make it something that is either passed in as an option or not:
```go
var ReadOnly optionFunc = func(r *Repo) {
	r.readonly = true
}
```
then we have to do some extra ceremony just for this option to work:
```go
	options := []git.Option{
		git.PollInterval(*gitPollInterval),
		git.Timeout(*gitTimeout),
	}
	if *gitReadonly {
		options = append(options git.ReadOnly())
	}

	repo := git.NewRepo(gitRemote, options...)
```
---

Due to that, I thought it best to follow the system the other options use so they're all uniform since it also has the benefit of being must more straightforward in actual usage now that the way this option is functionally utilized has changed.

Reply:
Having a uniform model of wrapping a single value is a red herring -- they already have a uniform model, that of implementing `interface { apply(*Repo) }`.

I'm not wedded to the way options are supplied. It is a bit weird to have an option that you can supply as switched off -- like saying `doSomething(..., optionalDifferentBehaviour(notDifferent))`. But using varargs is kind of a stylistic choice, and if it's more convenient across the two call sites to always supply the option, just with different values, I can abide by that.

Reply:
I changed the naming to `SetReadOnly` to make it more clear.  If that still presents an issue to you then please just make a clear and specific call on how you want it to look (i.e. option A or option B) and I will happily make it so.